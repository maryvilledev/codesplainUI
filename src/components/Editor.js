import React from 'react'
import CodeMirror from 'react-codemirror'
import { styleLine, styleAll, highlight } from '../util/codemirror-utils.js';
import { getTokenCount, getPrettyTokenName } from '../util/tokens.js';
import { parsePython3 } from '../parsers/python3';
import 'codemirror/lib/codemirror.css';
import 'codemirror/mode/python/python.js';
import '../styles/codesplain.css';

// Options for the CodeMirror instance that are shared by edit and annotation mddes
const baseCodeMirrorOptions = {
    lineNumbers: true,
    theme: 'codesplain',
    gutters: [ 'annotations', 'CodeMirror-linenumbers' ],
    language: 'python'
};

// Options specific for edit mode should be set here
const editModeOptions = {
  ...baseCodeMirrorOptions,
  readOnly: false,
};

// Options specific for annotation mode should be set here
const annotationModeOptions = {
  ...baseCodeMirrorOptions,
  readOnly: true,
  cursorBlinkRate: -1,
};

const makeMarker = () => {
  const marker = document.createElement("div");
  marker.style.color = "#822";
  marker.innerHTML = "●";
  return marker;
}

class Editor extends React.Component {
  constructor(props){
    super(props)
    this.state = {
      codeMirror: null
    }
    this.handleGutterClick = this.handleGutterClick.bind(this);
    this.emphasizeLine = this.emphasizeLine.bind(this);
    this.deEmphasize = this.deEmphasize.bind(this);
  }
  componentDidMount() {
    const codeMirror = this.codeMirror.getCodeMirror()
    codeMirror.on('gutterClick', this.handleGutterClick)
    this.startParserDaemon(parsePython3)
  }
  handleGutterClick(instance, line, gutter) {
    const { onGutterClick } = this.props
    const lineNumber = line
    const lineText = instance.getLine(line)
    onGutterClick(lineNumber, lineText)
  }
  // Can be used to emphasize a line of text when annotating.
  emphasizeLine(line) {
    const codeMirror = this.codeMirror.getCodeMirror()
    // Fade out the background
    const backgroundCSS = 'opacity: 0.5; font-weight: normal;';
    styleAll(codeMirror, backgroundCSS);

    // Bold the passed-in line
    const foregroundCSS = 'font-weight: bold; opacity: 1.0;';
    styleLine(codeMirror, line, foregroundCSS);
  }

  // If a line has been previously emphasized, this will de-emphasise it.
  deEmphasize() {
    const css = 'font-weight: normal; opacity: 1.0;';
    styleAll(this.codeMirror.getCodeMirror(), css);
  }
  // Runs the parser every second
  startParserDaemon(parser) {
    setInterval(function() {
      const snippet = this.props.value;
      if (snippet !== undefined && snippet !== '' && snippet !== this.state.prevSnippet) {
        // Generate an AST for the current state of the code snippet, if ready
        const AST = parser(snippet);

        // Get an array mapping token types to their occurence count
        const tokenCount = [];
        getTokenCount(AST, tokenCount);

        // Generate array of strings containing pretty token name and its count
        const filters = this.props.filters;
        let newFilters = {};
        Object.keys(tokenCount).filter(t => getPrettyTokenName(t) !== undefined)
          .forEach(t => {
            let selected = false;
            if (filters[t]) selected = filters[t].selected;
            newFilters[t] = {
              prettyTokenName: getPrettyTokenName(t),
              count: tokenCount[t],
              selected: selected,
            }
          });

        //Invoke prop callback
        this.setState({prevSnippet: snippet})
        this.props.onParserRun(AST, newFilters);
      }
      if(this.props.value) {
        highlight(this.codeMirror.getCodeMirror(), this.props.AST, this.props.filters);
      }
    }.bind(this), 1000);
  }
  render() {
    const {readOnly, onChange, markedLines, openLine, value} = this.props
    const codeMirrorOptions = (readOnly ? annotationModeOptions : editModeOptions)
    if (this.codeMirror) {
      const codeMirror = this.codeMirror.getCodeMirror()
      codeMirror.clearGutter('annotations')
      markedLines.forEach((line) => codeMirror.setGutterMarker(line, 'annotations', makeMarker()))
      this.deEmphasize();
      if (openLine !== undefined)
        this.emphasizeLine(openLine);
    }
    return (
      <div>
        <CodeMirror
          options={codeMirrorOptions}
          onChange={onChange}
          ref={(cm) => this.codeMirror = cm}
          value={value}
        />
      </div>
    )
  }
}

export default Editor
