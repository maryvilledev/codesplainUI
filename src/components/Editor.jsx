import React, { PropTypes } from 'react'
import CodeMirror from 'react-codemirror'
import {
  highlight,
  styleLine,
  styleAll,
} from '../util/codemirror-utils.js';
import { getRuleCount, rules } from '../util/rules.js';
import parsePython3 from '../parsers/python3.min.js';
import 'codemirror/lib/codemirror.css';
import 'codemirror/mode/python/python.js';
import '../styles/codesplain.css';

// Options for the CodeMirror instance that are shared by edit and annotation mddes
const baseCodeMirrorOptions = {
  lineNumbers: true,
  theme: 'codesplain',
  gutters: ['annotations', 'CodeMirror-linenumbers'],
  mode: 'python',
};

// Options specific for edit mode should be set here
const editModeOptions = {
  ...baseCodeMirrorOptions,
  readOnly: false,
};

// Options specific for annotation mode should be set here
const annotationModeOptions = {
  ...baseCodeMirrorOptions,
  readOnly: true,
  cursorBlinkRate: -1,
};

const makeMarker = () => {
  const marker = document.createElement('div');
  marker.style.color = '#822';
  marker.innerHTML = '●';
  return marker;
};

class Editor extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      prevSnippet: undefined,
    };
    this.handleGutterClick = this.handleGutterClick.bind(this);
    this.emphasizeLine = this.emphasizeLine.bind(this);
    this.deEmphasize = this.deEmphasize.bind(this);
    this.startParserDaemon = this.startParserDaemon.bind(this);
  }

  componentDidMount() {
    const codeMirror = this.codeMirror.getCodeMirror();
    codeMirror.on('gutterClick', this.handleGutterClick);
    this.startParserDaemon(parsePython3);
  }

  componentDidUpdate() {
    const {
      markedLines,
      openLine,
    } = this.props;

    const codeMirrorInst = this.codeMirror.getCodeMirror();
    codeMirrorInst.clearGutter('annotations');
    // eslint-disable-next-line array-callback-return
    markedLines.forEach((lineNumber) => {
      codeMirrorInst.setGutterMarker(Number(lineNumber), 'annotations', makeMarker());
    });
    this.deEmphasize();
    if (openLine) {
      this.emphasizeLine(openLine);
    }
  }

  handleGutterClick(instance, lineNumber) {
    const { onGutterClick } = this.props;
    const lineText = instance.getLine(lineNumber);
    onGutterClick(lineNumber, lineText);
  }

  // Can be used to emphasize a line of text when annotating.
  emphasizeLine(line) {
    const codeMirror = this.codeMirror.getCodeMirror();
    // Fade out the background
    const backgroundCSS = 'opacity: 0.5; font-weight: normal;';
    styleAll(codeMirror, backgroundCSS);

    // Bold the passed-in line
    const foregroundCSS = 'font-weight: bold; opacity: 1.0;';
    styleLine(codeMirror, line, foregroundCSS);
  }

  // If a line has been previously emphasized, this will de-emphasise it.
  deEmphasize() {
    const css = 'font-weight: normal; opacity: 1.0;';
    styleAll(this.codeMirror.getCodeMirror(), css);
  }

  // Runs the parser every second
  startParserDaemon(parser) {
    setInterval(() => {
      const snippet = this.props.value;
      if (snippet && snippet !== this.state.prevSnippet) {
        // Generate an AST for the current state of the code snippet, if ready
        const onError = (err) => { console.log(err) };
        const AST = parser(snippet, onError);

        // Get an array mapping token types to their occurence count
        const ruleCount = {};
        getRuleCount(AST, ruleCount);

        // Generate array of strings containing pretty token name and its count
       const filters = this.props.filters;
       let newFilters = {};
       Object.keys(ruleCount).filter(r => rules[r] !== undefined)
         .forEach(r => {
           let selected = false;
           if (filters[r]) {
             selected = filters[r].selected;
           }
           newFilters[r] = {
             prettyTokenName: rules[r].prettyName,
             count:           ruleCount[r],
             selected:        selected,
           }
         });

        // Invoke prop callback
        this.setState({ prevSnippet: snippet });
        this.props.onParserRun(AST, newFilters);
      }

      if (this.props.value) {
        highlight(this.codeMirror.getCodeMirror(), this.props.AST, this.props.filters);
      }
    }, 1000);
  }

  render() {
    const {
      onChange,
      readOnly,
      value,
    } = this.props;

    const codeMirrorOptions = {
      ...(readOnly ? annotationModeOptions : editModeOptions),
    };

    return (
      <div>
        <CodeMirror
          onChange={onChange}
          options={codeMirrorOptions}
          ref={cm => this.codeMirror = cm}
          value={value}
        />
      </div>
    );
  }
}

Editor.propTypes = {
  AST: PropTypes.object.isRequired,
  filters: PropTypes.object.isRequired,
  markedLines: PropTypes.arrayOf(PropTypes.number).isRequired,
  onChange: PropTypes.func.isRequired,
  onGutterClick: PropTypes.func.isRequired,
  onParserRun: PropTypes.func.isRequired,
  openLine: PropTypes.number,
  readOnly: PropTypes.bool.isRequired,
  value: PropTypes.string.isRequired,
};

export default Editor;
