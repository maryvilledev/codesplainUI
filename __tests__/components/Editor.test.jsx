import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import Editor from '../../src/components/Editor';

const defaultProps = {
  AST: {},
  filters: {},
  language: 'python3',
  markedLines: [],
  onChange: jest.fn(),
  onGutterClick: jest.fn(),
  readOnly: false,
  value: '',
};

describe('<Editor />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = (node) => shallow(node, { context: { muiTheme } });

  it('matches snapshot', () => {
    const wrapper = shallowWithContext(
      <Editor {...defaultProps} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('prop: onChange', () => {
    it('forwarded to the CodeMirror instance', () => {
      const onChange = jest.fn();
      const wrapper = shallowWithContext(
        <Editor
          {...defaultProps}
          onChange={onChange}
        />
      );
      expect(wrapper.find('CodeMirror').prop('onChange')).toEqual(onChange);
    });
  });

  describe('prop: value', () => {
    it('forwarded to the CodeMirror instance', () => {
      const snippetContents = "What up, my glib globs!";
      const wrapper = shallowWithContext(
        <Editor
          {...defaultProps}
          value={snippetContents}
        />
      );
      expect(wrapper.find('CodeMirror').prop('value')).toEqual(snippetContents);
    });
  });
});
