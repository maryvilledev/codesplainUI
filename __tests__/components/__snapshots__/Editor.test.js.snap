exports[`<Editor /> matches snapshot 1`] = `
<div>
  <CodeMirror
    onChange={[Function]}
    options={
      Object {
        "gutters": Array [
          "annotations",
          "CodeMirror-linenumbers",
        ],
        "lineNumbers": true,
        "mode": "python",
        "readOnly": false,
        "theme": "codesplain",
      }
    }
    preserveScrollPosition={false}
    value="" />
</div>
`;
